modules/idLibrary.js

```diff
@@ -0,0 +1,238 @@
+  const matched = value.match(/([a-zA-Z0-9._-]+@[a-zA-Z0-9._-]+\.[a-zA-Z0-9._-]+)/gi);
+  if (!matched) {
+    return null;
+  }
+  logInfo('Email found' + matched[0]);
+  return matched[0];
+}
+
+function bodyAction(conf, mutations, observer) {
+  logInfo('BODY observer on debounce called');
+
+  if (email) {
+    observer.disconnect();
+    logInfo('Email is found, body observer disconnected');
+  }
+
+  const body = document.body.innerHTML;
+  email = getEmail(body);
+  if (email != null) {
+    logInfo(`Email obtained from the body ${email}`);
+    observer.disconnect();
+    logInfo('Post data on email found in body');
+    postData(conf.url);
+  }
+}
+
+function targetAction(conf, mutations, observer) {
+  logInfo('Target observer called');
+  for (const mutation of mutations) {
+    for (const node of mutation.addedNodes) {
+      email = node.textContent;
+
+      if (email) {
+        logInfo('Email obtained from the target ' + email);
+        observer.disconnect();
+        logInfo(' Post data on email found in target');
+        postData(conf.url);
+      }
+    }
+  }
+}
+
+function addInputElementsElementListner(conf) {
+  logInfo('Adding input element listeners');
+  const inputs = document.querySelectorAll('input[type=text], input[type=email]');
+
+  for (var i = 0; i < inputs.length; i++) {
+    logInfo(` Original Value in Input = ${inputs[i].value}`);
+    inputs[i].addEventListener('change', event => processInputChange(event, conf));
+    inputs[i].addEventListener('blur', event => processInputChange(event, conf));
+  }
+}
+
+function removeInputElementsElementListner(conf) {
+  logInfo('Removing input element listeners');
+  const inputs = document.querySelectorAll('input[type=text], input[type=email]');
+
+  for (var i = 0; i < inputs.length; i++) {
+    inputs[i].removeEventListener('change', event => processInputChange(event, conf));
+    inputs[i].removeEventListener('blur', event => processInputChange(event, conf));
+  }
+}
+
+function processInputChange(event, conf) {
+  const value = event.target.value;
+  logInfo(`Modified Value of input ${event.target.value}`);
+  email = getEmail(value);
+  if (email != null) {
+    logInfo('Email found in input ' + email);
+    postData(conf.url);
+    removeInputElementsElementListner(conf);
+  }
+}
+
+function debounce(func, wait, immediate) {
+  let timeout;
+
+  return function (...args) {
+    const context = this;
+    const later = function () {
+      timeout = null;
+      if (!immediate) {
+        func.apply(context, args);
+      }
+    };
+    const callNow = immediate && !timeout;
+    clearTimeout(timeout);
+    timeout = setTimeout(later, wait);
+    if (callNow) {
+      func.apply(context, args);
+    }
+  };
+}
+
+function handleTargetElement(conf) {
+  const targetObserver = new MutationObserver(function (mutations, observer) {
+    logInfo('target observer called');
+    debounce(targetAction(conf, mutations, observer), conf.debounce, false);
```

Comment:
@SKOCHERI 

What is the intent here? Please elaborate on the expectation of things to happen when this line is hit.

`debounce` takes in a `func` param that looks like the intent is for it to be called later.

But you are executing the `targetAction` here with those params, which returns undefined. So really `func` is always undefined when `debounce` is run.

Is the intent to not call `targetAction` here but just pass its reference to debounce?

If so, recommend you use `bind`.

```js
debounce(targetAction.bind(null, conf, mutations, observer), conf.debounce, false);
```

Reply:
The debounce issue is fixed